pyramid/directory.go

```diff
@@ -0,0 +1,73 @@
+package pyramid
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"sync"
+)
+
+// directory allows synchronization between file operation that might
+// change (create/delete) directories
+type directory struct {
+	sync.Mutex
+}
+
+func (d *directory) deleteDirRecIfEmpty(dir string) error {
+	d.Lock()
+	defer d.Unlock()
+
+	for {
+		empty, err := isDirEmpty(dir)
+		if err != nil {
+			return err
+		}
+		if !empty {
+			return nil
+		}
+
+		parentDir := path.Dir(dir)
+		if err := os.Remove(dir); err != nil {
+			return err
+		}
+		dir = parentDir
```

Comment:
This loop can continue _way_ up, above the planned location.  I think `directory` needs a concept of a `ceilingDir` or `firstAncestorDir` or something.

Reply:
I was planning for rooted `workspace` dir to back me up, but it's indeed flaky. Fixed.

Reply:
Adding documentation for now, really can't see how the paths (other than the base directory) will be user-controlled in the future.

Reply:
Cool.  I don't particularly mind "user-controlled" paths when that user is the owner of lakeFS.  I'm worried about "user-controlled" paths when they come from a lakeFS user, who is supposed only to have permissions to act on files _inside_ lakeFS.
Thanks!